javascript/ql/src/semmle/javascript/frameworks/EventEmitter.qll

```diff
@@ -0,0 +1,211 @@
+import javascript
+
+module EventEmitter {
+  /** Gets the name of a method on `EventEmitter` that returns `this`. */
+  string chainableMethod() {
+    result = "off" or
+    result = "removeAllListeners" or
+    result = "removeListener" or
+    result = "setMaxListeners" or
+    result = on()
+  }
+
+  /** Gets the name of a method on `EventEmitter` that registers an event handler. */
+  string on() {
+    result = "addListener" or
+    result = "on" or
+    result = "once" or
+    result = "prependListener" or
+    result = "prependOnceListener"
+  }
+
+  /**
+   * An instance of the NodeJS EventEmitter class.
+   * Extend this class to mark something as being an instance of the EventEmitter class.
+   */
+  final class EventEmitter extends DataFlow::Node {
+    EventEmitterRange::Range range;
+
+    EventEmitter() { this = range }
+
+    /**
+     * Get a method name that returns `this` on this type of emitter.
+     */
+    string getAChainableMethod() { result = range.getAChainableMethod() }
+
+    /**
+     * Get a reference through type-tracking to this EventEmitter.
+     * The type-tracking tracks through chainable methods.
+     */
+    DataFlow::SourceNode ref() { result = range.ref() }
+  }
+
+  module EventEmitterRange {
+    abstract class Range extends DataFlow::Node {
+      string getAChainableMethod() { result = EventEmitter::chainableMethod() }
+
+      private DataFlow::SourceNode ref(DataFlow::TypeTracker t) {
```

Comment:
I've since changed it.  

`EventEmitter::Range` no longer implementes `.ref()`, as it is not directly needed in that class any more. (I had issues with non-monotonic recursion when trying to implement SocketIO with that design).    
Altough currently every concrete subclass happens to implement a `.ref()` method.